numpy/_core/tests/test_multiarray.py

```diff
@@ -1984,6 +1984,14 @@ def test_choose(self):
         y = np.choose([0, 0, 0], [x[:3], x[:3], x[:3]], out=x[1:4], mode='wrap')
         assert_equal(y, np.array([0, 1, 2]))
 
+        # gh-28206 test new StringDType casting
+        x = np.array(['a', 'b', 'c'], dtype=np.dtypes.StringDType())
+        y = np.array(['abc', 'def', 'ghi'], dtype=np.dtypes.StringDType())
```

Comment:
Subtle thing, but make sure to include a long string (otherwise you can't trigger the path which might have crashed).

Reply:
Actually that still seems to crash, thanks for noticing. I think `PyArray_ConvertToCommonType` (it seems to be a backward compatibility function) does not cast properly to a common scenario for `StringDType`, and `EquivTypes` passes the `dtype` unchanged--so, we end up with the `dtype` only for the first array. Should we do a promotion of some sort if the types are equivalent? Or perhaps, for now, we can ensure they are _identical_?